rust/feature-flags/src/flag_matching.rs

```diff
@@ -805,6 +828,37 @@ impl FeatureFlagMatcher {
         }
     }
 
+    /// Evaluates dynamic cohort property filters
+    ///
+    /// NB: This method first caches all of the cohorts associated with the team, which allows us to avoid
+    /// hitting the database for each cohort filter.
+    pub async fn evaluate_cohort_filters(
+        &self,
+        cohort_property_filters: &[PropertyFilter],
+        target_properties: &HashMap<String, Value>,
+    ) -> Result<bool, FlagError> {
+        // At the start of the request, fetch all of the cohorts for the team from the cache
+        // This method also caches the cohorts in memory for the duration of the application, so we don't need to fetch from
+        // the database again until we restart the application.
+        let cohorts = self.cohort_cache.get_cohorts_for_team(self.team_id).await?;
+
+        // Store cohort match results in a HashMap to avoid re-evaluating the same cohort multiple times,
+        // since the same cohort could appear in multiple property filters. This is especially important
+        // because evaluating a cohort requires evaluating all of its dependencies, which can be expensive.
+        let mut cohort_matches = HashMap::new();
+        for filter in cohort_property_filters {
+            let cohort_id = filter
+                .get_cohort_id()
+                .ok_or(FlagError::CohortFiltersParsingError)?;
+            let match_result =
+                evaluate_cohort_dependencies(cohort_id, target_properties, cohorts.clone())?;
```

Comment:
This /definitely/ doesn't need a clone of the cohorts vec. Cohorts are cheap, but not free, cloning the cohort set once per request is already expensive enough. This diff compiles:
```diff
@@ -1108,10 +1108,9 @@ impl FeatureFlagMatcher {
 fn evaluate_cohort_dependencies(
     initial_cohort_id: CohortId,
     target_properties: &HashMap<String, Value>,
-    cohorts: Vec<Cohort>,
+    cohorts: &[Cohort],
 ) -> Result<bool, FlagError> {
-    let cohort_dependency_graph =
-        build_cohort_dependency_graph(initial_cohort_id, cohorts.clone())?;
+    let cohort_dependency_graph = build_cohort_dependency_graph(initial_cohort_id, cohorts)?;

     // We need to sort cohorts topologically to ensure we evaluate dependencies before the cohorts that depend on them.
     // For example, if cohort A depends on cohort B, we need to evaluate B first to know if A matches.
@@ -1216,7 +1215,7 @@ fn apply_cohort_membership_logic(
 /// The graph is acyclic, which is required for valid cohort dependencies.
 fn build_cohort_dependency_graph(
     initial_cohort_id: CohortId,
-    cohorts: Vec<Cohort>,
+    cohorts: &[Cohort],
 ) -> Result<DiGraph<CohortId, ()>, FlagError> {
     let mut graph = DiGraph::new();
     let mut node_map = HashMap::new();
```